core/autoscaling.go

```diff
@@ -505,89 +647,299 @@ func (a *autoScalingGroup) getAnySpotInstance() *instance {
+			return InstanceStateCodeTerminated
+		}
+		logger.Println("Error describing instance status:", *instanceID, err)
+		return InstanceStateCodePending
+	}
+
+	if len(out.InstanceStatuses) > 0 {
+		return *out.InstanceStatuses[0].InstanceState.Code
+	}
+
+	// If instance is not showing then it's terminated
+	return InstanceStateCodeTerminated
+}
+
+func canTerminateInstance(instanceID *string, instanceState int64) bool {
+	return instanceID != nil && instanceState != InstanceStateCodeTerminated && instanceState != InstanceStateCodeShuttingDown
+}
+
+func (a *autoScalingGroup) terminateInstance(instanceID *string, instanceState int64) bool {
+	svc := a.region.services.ec2
+	if canTerminateInstance(instanceID, instanceState) {
+		_, err := svc.TerminateInstances(&ec2.TerminateInstancesInput{
+			InstanceIds: []*string{instanceID},
+		})
+		if err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (a *autoScalingGroup) cancelSIR(sirRequestID *string) error {
+	svc := a.region.services.ec2
+	_, err := svc.CancelSpotInstanceRequests(
+		&ec2.CancelSpotInstanceRequestsInput{
+			SpotInstanceRequestIds: []*string{sirRequestID},
+		})
+
+	return err
+}
+
+func (a *autoScalingGroup) cancelSIRAndTerminateInstance(instanceID *string, sirRequestID *string, instanceState int64) {
+
+	if !a.terminateInstance(instanceID, instanceState) {
+		logger.Println("Unable to terminate instance:", *instanceID)
+		return
+	}
+
+	err := a.cancelSIR(sirRequestID)
+
+	if err == nil {
+		a.markSpotInstanceRequestAsCompete(sirRequestID)
+	}
+}
+
+func (a *autoScalingGroup) processUnattachedRunningInstance(req *spotInstanceRequest, instanceID *string) (bool, bool) {
+	logger.Println(a.name, "Active bid was found, with running "+
+		"instances not yet attached to the ASG",
+		*instanceID)
+	// we need to re-scan in order to have the information a
+	err := req.region.scanInstances()
+	if err != nil {
+		logger.Printf("Failed to scan instances: %s for %s\n", err, req.asg.name)
+	}
+
+	tags := req.asg.propagatedInstanceTags()
+
+	i := req.region.instances.get(*instanceID)
+
+	if i != nil {
+		i.tag(tags, defaultTimeout)
+	} else {
+		logger.Println(req.asg.name, "new spot instance", *instanceID, "has disappeared")
+	}
+	return true, false
+}
+
+func isInstanceRunning(spotInstanceRunning int64) bool {
+	return spotInstanceRunning == InstanceStateCodeRunning
+}
+
+func isInstanceInUsableState(spotInstanceRequest int64) bool {
+	return spotInstanceRequest > InstanceStateCodeRunning
+}
+
+func (a *autoScalingGroup) processUnattachedInstance(req *spotInstanceRequest, instanceID *string) (bool, bool) {
+	// var validRequest *spotInstanceRequest
+	// processNextSIR, waitForNextExecution := false, false
+
+	spotInstanceRunning := a.getInstanceState(instanceID)
+	if isInstanceRunning(spotInstanceRunning) {
+		return a.processUnattachedRunningInstance(req, instanceID)
+	} else if isInstanceInUsableState(spotInstanceRunning) {
+		// stopped, terminated, shutting-dowm etc,
```

Comment:
typo: dowm

Reply:
fixed this typo.  It appears life imitates art; and I'm unable to spell complete, as I am able to complete things :-)